test/index.html

```diff
@@ -0,0 +1,247 @@
+                         next : function(){ return impress().next(); },
+                         text : "next() wraps around to first step (4->1)" },
+                       { me   : step1,
+                         next : function(){ return impress().prev(); },
+                         text : "prev() wraps around to last step (1->4)" },
+                       { me   : step4,
+                         next : function(){ return impress().prev(); },
+                         text : "prev() called and returns ok (4->3)" },
+                       { me   : step3,
+                         next : false }      
+      ];
+      
+      var assertStepEnter = function( event, registeredListener ) {
+        assert.equal( event.target, nextStep[i].me,
+                      event.target.id + " triggered impress:stepenter event." );
+        assert.ok( event.target.classList.contains("active"),
+                   event.target.id + " set active css class." );
+        assert.ok( event.target.classList.contains("present"),
+                   event.target.id + " set present css class." );
+        assert.ok( !event.target.classList.contains("future"),
+                   event.target.id + " unset future css class." );
+        assert.ok( !event.target.classList.contains("past"),
+                   event.target.id + " unset past css class." );
+
+        if( nextStep[i].next ) {
+          assert.ok( nextStep[i].next(), nextStep[i].text );
+          i++;
+        } else {
+          // Remember to cleanup, since we're operating outside of qunit-fixture
+          root.removeEventListener( "impress:stepenter", assertStepEnterWrapper );
+          doneWithNext();
+        }
+      };
+      
+      var assertStepEnterWrapper = function( event ) {
+        setTimeout( function() { assertStepEnter( event ) },
+                    wait*2 ); // wait*2 ensures that assertStepLeave is finished 
+                              // first, since here we'll call nextStep.next and i++
+      };
+      root.addEventListener( "impress:stepenter", assertStepEnterWrapper );
+
+
+      var prevStep = [ step1 ];
+      for(j=0; j < nextStep.length; j++){
+        prevStep.push( nextStep[j].me );
+      }
+      var assertStepLeave = function( event, registeredListener ) {
+        assert.equal( event.target, prevStep[i],
+                      event.target.id + " triggered impress:stepleave event." );
+        assert.ok( !event.target.classList.contains("active"),
+                   event.target.id + " unset active css class." );
+        assert.ok( !event.target.classList.contains("present"),
+                   event.target.id + " unset present css class." );
+        assert.ok( !event.target.classList.contains("future"),
+                   event.target.id + " unset future css class." );
+        assert.ok( event.target.classList.contains("past"),
+                   event.target.id + " set past css class." );
+        
+        if( i+2 == prevStep.length ) {
```

Comment:
What is easier to read is also easier to maintain. Having something that is easier to maintain but hard to read seems to be a contradiction.

I understand your point about test execution time, the problem is that if you have too many premisses in a single test then it will be more dificult to create a deterministic assertion in the future when you still have to maintain the previous ones. Having a single atomic block of code that test a specific path turns out to be much more maintainable and scalable when developing regression tests.

What I suggest is that, instead of creating a test called "Impress core API", it should instead be created a "module" called "Impress core API" with tests that cover several situations separately. This way if one wants to run the tests that tests for only a specific event or animation flow, it is just clicking in the desired test in QUnit UI that it will go straight to it, without having to worry about the others. That is very useful for debugging purposes.

Below there are some examples:

**Found the failing test among others**

![image](https://cloud.githubusercontent.com/assets/835857/12886736/9b128ab2-cec3-11e5-8897-1f8cae80c259.png)

**Clicked on it to see why it is failing**

![image](https://cloud.githubusercontent.com/assets/835857/12886757/b018d218-cec3-11e5-9a9a-0a6fb533653c.png)